### PR TITLE
Fix EditorConfig at solution level

### PR DESCRIPTION
**Pull request by beccamc**
Update the editor config creation flow to use the external access point provided by Roslyn. 

templates_working

Also fixing this issue - https://github.com/dotnet/templates/issues/718

### CODE REVIEW

**Comment by beccamc** [comment]
@jasonmalinowski Could you take a look at this when you get a chance? I think you have the most context here and I'd love another set of eyes.

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs [hunk 1]
```diff
@@ -1,56 +1,22 @@
...
 public class RoslynEditorConfigFileGenerator
 {
-    private const BindingFlags PublicBindingFlags = BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public;
-    private const BindingFlags NonPublicBindingFlags = BindingFlags.Instance | BindingFlags.Static | BindingFlags.NonPublic;
-
     public RoslynEditorConfigFileGenerator(DTE2 dte)
     {
         try
         {
-            string? basePath = Path.GetDirectoryName(dte.FullName);
```
Comment (jasonmalinowski) [inline src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs:30]: Great to see so much being deleted here. And wow this code was scary and broken...

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs [hunk 2]
```diff
@@ -1,56 +1,22 @@
...
-            string languageServiceSubPath = @"CommonExtensions\Microsoft\VBCSharp\LanguageServices";
-            var baseDirectory = Path.Combine(basePath!, languageServiceSubPath);
-
-            _microsoftVisualStudioLanguageServicesVisualBasicAssembly =
-                new Lazy<Assembly>(() => Assembly.LoadFrom(Path.Combine(baseDirectory, "Microsoft.VisualStudio.LanguageServices.VisualBasic.dll")));
-
-            _microsoftVisualStudioLanguageServicesCSharpAssembly =
-                new Lazy<Assembly>(() => Assembly.LoadFrom(Path.Combine(baseDirectory, "Microsoft.VisualStudio.LanguageServices.CSharp.dll")));
-
             _serviceProvider = new ServiceProvider(dte as OLE.Interop.IServiceProvider);
```
Comment (jasonmalinowski) [inline src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs:19]: It's a bit strange here that this is using DTE as the service provider. I'm not sure why this isn't just using https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.serviceprovider.globalprovider?view=visualstudiosdk-2022 or similar. And at that point you can just delete this constructor entirely.

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs [hunk 3]
```diff
@@ -1,56 +1,22 @@
...
-                typeof(ImmutableArray<(string feature, ImmutableArray<IOption> options)>),
-                typeof(OptionSet),
-                typeof(string)
-                };
-
-                return editorConfigFileGeneratorType?.GetMethod("Generate", PublicBindingFlags, binder: null, types: targetParams, modifiers: null);
-            });
         }
         catch (Exception ex)
         {
```
Comment (jasonmalinowski) [inline src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs:22]: "Roslyn" has a typo on the line below this. Not your change but if you're here...(although we can really just delete this...)

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs [hunk 4]
```diff
@@ -61,185 +27,18 @@ public RoslynEditorConfigFileGenerator(DTE2 dte)
...
-    }
-
-    private string? Generate(ImmutableArray<(string feature, ImmutableArray<IOption> options)> groupedOptions, OptionSet optionSet, string language)
+    public string? Generate(string language)
     {
         try
         {
-            LogEvents(EventId.Options, new OptionsInfo(groupedOptions, optionSet));
-            return (string?)_generatorMethod.Value?.Invoke(null, new object[] { groupedOptions, optionSet, language });
+            var editorConfigGenerator = _serviceProvider.GetService(typeof(IEditorConfigGenerator)) as IEditorConfigGenerator;
```
Comment (jasonmalinowski) [inline src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs:35]:
```suggestion
            var editorConfigGenerator = (IEditorConfigGenerator)_serviceProvider.GetService(typeof(IEditorConfigGenerator));
```

The "as" cast means this becomes undebuggable if it ever fails.

Comment (jasonmalinowski) [inline src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs:35]: I'm surprised this is using service providers here given this line:

https://github.com/dotnet/roslyn/blob/caa98ccbdd7947397506d62bee5d6dd7b4e95b5a/src/Tools/ExternalAccess/EditorConfigGenerator/EditorConfigGenerator.cs#L15

Is using MEF to export. I'd expect this to always fail -- is it?

File: eng/Versions.targets [hunk 5]
```diff
@@ -4,6 +4,7 @@
     <PackageReference Update="EnvDTE" Version="$(EnvDTEVersion)" />
     <PackageReference Update="EnvDTE80" Version="$(EnvDTE80Version)" />
     <PackageReference Update="Microsoft.CodeAnalysis.Workspaces.Common" Version="$(MicrosoftCodeAnalysisWorkspacesCommonVersion)" />
+    <PackageReference Update="Microsoft.CodeAnalysis.ExternalAccess.EditorConfigGenerator" Version="$(MicrosoftCodeAnalysisWorkspacesCommonVersion)" />
```
Comment (jasonmalinowski) [inline eng/Versions.targets:7]: Do you want to be defining a new variable for the EditorConfigGenerator package? Or just consider using CommonMicrosoftCodeAnalysisVersion?

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Microsoft.VisualStudio.Templates.Editorconfig.Wizard.csproj [hunk 6]
```diff
@@ -11,11 +11,13 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis.Workspaces.Common" />
     <PackageReference Include="Microsoft.VisualStudio.LanguageServices" />
-    <PackageReference Include="Microsoft.VisualStudio.SDK" />
-    <PackageReference Include="Microsoft.VisualStudio.Telemetry" />
-    <PackageReference Include="Microsoft.VisualStudio.TemplateWizardInterface" />
-    <PackageReference Include="StreamJsonRpc" />
+    <PackageReference Include="Microsoft.CodeAnalysis.ExternalAccess.EditorConfigGenerator" />
+    <PackageReference Include="Microsoft.VisualStudio.SDK" Version="17.9.35087-preview.1" />
```
Comment (jasonmalinowski) [inline src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Microsoft.VisualStudio.Templates.Editorconfig.Wizard.csproj:15]: It's a bit odd we're mixing central package versions and inline versions here. Did we want to standardize to one?

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Utilities/VSHelpers.cs [hunk 7]
```diff
@@ -78,6 +78,13 @@ public static (bool isSolutionLevel, string? path, string? language, object? sel
         foreach (UIHierarchyItem selItem in items)
         {
             selectedItem = selItem.Object;
+
+            // Check if selected item is the solution 
+            if (selectedItem is EnvDTE.Solution)
+            {
+                return (true, Path.GetDirectoryName(DTE.Solution.FullName), HasVisualBasicProjects() ? LanguageNames.VisualBasic : LanguageNames.CSharp, selectedItem);
```
Comment (jasonmalinowski) [inline src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Utilities/VSHelpers.cs:85]: What happens if there's both C# and VB projects?

Comment (beccamc) [reply src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Utilities/VSHelpers.cs:85]: In 2019 where this feature still  works, it just generates for C#. I am not going to change any functionality without customer input. 

image

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs (hunk 4 above)
Comment (beccamc) [reply src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs]: Thank you!

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Microsoft.VisualStudio.Templates.Editorconfig.Wizard.csproj (hunk 6 above)
Comment (beccamc) [reply src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Microsoft.VisualStudio.Templates.Editorconfig.Wizard.csproj]: I moved over to standardize. The system had some issues with the RPC package versioning so I'm going to leave those inline.

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs (hunk 4 above)
Comment (beccamc) [reply src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs]: Okay, I think this should work now?

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs (hunk 2 above)
Comment (beccamc) [reply src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs]: Updated to the GlobalProvider

**Comment by arkalyanms** [comment]
Retriggered the expired gates. Is this ready to go in?

File: src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs [hunk 8]
```diff
@@ -1,245 +1,32 @@
...
     }
 
-    private string? Generate(ImmutableArray<(string feature, ImmutableArray<IOption> options)> groupedOptions, OptionSet optionSet, string language)
+    public string? Generate(string language)
     {
         try
         {
-            LogEvents(EventId.Options, new OptionsInfo(groupedOptions, optionSet));
-            return (string?)_generatorMethod.Value?.Invoke(null, new object[] { groupedOptions, optionSet, language });
+            var componentModel = (IComponentModel)ServiceProvider.GlobalProvider.GetService(typeof(SComponentModel));
```
Comment (beccamc) [inline src/Tools/Microsoft.VisualStudio.Templates.Editorconfig.Wizard/Generator/RoslynEditorConfigFileGenerator.cs:22]: updated service here

**Comment by beccamc** [comment]
Now waiting on PR https://github.com/dotnet/roslyn/pull/71003

**Comment by beccamc** [comment]
This required the change from https://github.com/dotnet/roslyn/pull/71003. I manually tested with that change applied to VS, and the ExternalAccess point worked. This PR can go in.

**Pull request merged**